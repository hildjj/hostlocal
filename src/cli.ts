#!/usr/bin/env node

import {
  Command,
  InvalidArgumentError,
  Option,
  OutputConfiguration,
} from 'commander';
import {DEFAULT_HOST_OPTIONS} from './opts.js';
import {hostLocal} from './index.js';
import {version} from './version.js';

function appendArray(val: string, prev: string[] | undefined): string[] {
  prev ??= [];
  prev.push(val);
  return prev;
}

function toInt(val: string): number {
  const ret = parseInt(val, 10);
  if (isNaN(ret)) {
    throw new InvalidArgumentError('not a valid integer.');
  }
  return ret;
}

/**
 * Run the CLI for the given arguments.
 *
 * @param args Arguments to process, in node format.  If not specified, uses
 *   process.argv.
 * @param out Override stdout and stderr for testing.
 */
export async function cli(
  args?: string[],
  out?: OutputConfiguration
): Promise<void> {
  const program = new Command();
  if (out) {
    program.configureOutput(out);
    program.exitOverride();
  }

  // Set default docs manually.
  await program
    .version(version)
    .option('-6, --ipv6', 'Listen on IPv6 only, if host supports both IPv4 and IPv6.')
    .option('-c, --config <file>', `If the given file exists, import it as a module and use its default export as the options.  Name is relative to cwd. Command line parameters overwrite options from the config file. (default: "${DEFAULT_HOST_OPTIONS.config}")`)
    .option('--certDir <directory>', `Directory, relative to cwd, to cache cert info. (default: "${DEFAULT_HOST_OPTIONS.certDir}")`)
    .option('-g, --glob <pattern>', 'Set of files to watch.  If one of these changes, execute the command in the --exec option.  Can be specified multiple times.', appendArray)
    .option('-e, --exec <shell command>', `Execute this command when the glob changes. (default: "${DEFAULT_HOST_OPTIONS.exec}")`)
    .option('-H, --host <address>', 'Hostname or IP address to listen on. "::" for everything. (default: "localhost")')
    .option('-i, --initial', 'If glob is specified, run the exec command on startup, before listening')
    .option('--notAfterDays <number>', `How many days is the certificate valid? (default: ${DEFAULT_HOST_OPTIONS.notAfterDays})`, toInt)
    .option('-o, --open <path>', `Open this path in the default browser.  Relative to server root and prefix, if specified.  If empty (""), do not open anything. (default: "${DEFAULT_HOST_OPTIONS.open}")`)
    .option('-p, --port <number>', `Port to serve content from.  Use 0 to get an unused port. (default: ${DEFAULT_HOST_OPTIONS.port})`, toInt)
    .option('-P, --prefix <string>', 'Make all of the URLs served have paths that start with this prefix, followed by a slash.')
    .option('-q, --quiet', 'Do not do logging')
    .option('--rawMarkdown', 'Do not process markdown into HTML')
    .option('-t, --timeout <number>', 'Time, in ms, to allow exec to run.', toInt)
    .addOption(
      // Testing only
      new Option('--shutTimes <number>')
        .argParser(toInt)
        .hideHelp()
    )
    .argument('[directory]', 'Directory to serve. (default: cwd)')
    .configureHelp({
      sortOptions: true,
    })
    .action((directory, opts) => hostLocal(directory, opts))
    .parseAsync(args);
}
