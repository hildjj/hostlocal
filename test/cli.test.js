import assert from 'node:assert';
import {cli} from '../lib/cli.js';
import snap from 'snappy-snaps';
import test from 'node:test';

test('help', async() => {
  let stdout = '';
  let stderr = '';
  const out = {
    writeOut(str) {
      stdout += str;
    },
    writeErr(str) {
      stderr += str;
    },
    outputError(str, write) {
      write(str);
    },
  };
  await assert.rejects(() => cli(['node', 'hostlocal', '-h'], out));
  assert.deepEqual(stdout, await snap('help stdout', stdout));
  assert.deepEqual('', await snap('help stderr', stderr));

  // eslint-disable-next-line require-atomic-updates
  stdout = '';
  // eslint-disable-next-line require-atomic-updates
  stderr = '';
  await assert.rejects(() => cli(['node', 'hostlocal', '-p', 'aaaa'], out));
  assert.deepEqual('', await snap('parse stdout', stdout));
  assert.deepEqual(stderr, await snap('parse stderr', stderr));
});
