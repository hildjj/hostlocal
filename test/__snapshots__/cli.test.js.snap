// Data Snap v1

exports[`help stderr`] = {
  "data": "",
  "meta": {
    "expires": null
  }
}

exports[`help stdout`] = {
  "data": "Usage: hostlocal [options] [directory]\n\nArguments:\n  directory                   Directory to serve. (default: cwd)\n\nOptions:\n  -6, --ipv6                  Listen on IPv6 only, if host supports both IPv4\n                              and IPv6.\n  -c, --config \u003Cfile\u003E         If the given file exists, import it as a module\n                              and use its default export as the options.  Name\n                              is relative to cwd. Command line parameters\n                              overwrite options from the config file. (default:\n                              \".hostlocal.js\")\n  --certDir \u003Cdirectory\u003E       Directory, relative to cwd, to cache cert info.\n                              (default: \".cert\")\n  -e, --exec \u003Cshell command\u003E  Execute this command when the glob changes.\n                              (default: \"npm run build\")\n  -g, --glob \u003Cpattern\u003E        Set of files to watch.  If one of these changes,\n                              execute the command in the --exec option.  Can be\n                              specified multiple times.\n  -h, --help                  display help for command\n  -H, --host \u003Caddress\u003E        Hostname or IP address to listen on. \"::\" for\n                              everything. (default: \"localhost\")\n  -i, --initial               If glob is specified, run the exec command on\n                              startup, before listening\n  --notAfterDays \u003Cnumber\u003E     How many days is the certificate valid? (default:\n                              7)\n  -o, --open \u003Cpath\u003E           Open this path in the default browser.  Relative\n                              to server root.  If empty, do not open anything.\n                              (default: \"\u002F\")\n  -p, --port \u003Cnumber\u003E         Port to serve content from.  Use 0 to get an\n                              unused port. (default: 8111)\n  -P, --prefix \u003Cstring\u003E       Make all of the URLs served have paths that start\n                              with this prefix, followed by a slash.\n  -q, --quiet                 Do not do logging\n  --rawMarkdown               Do not process markdown into HTML\n  -t, --timeout \u003Cnumber\u003E      Time, in ms, to allow exec to run.\n  -V, --version               output the version number\n",
  "meta": {
    "expires": null
  }
}

exports[`parse stderr`] = {
  "data": "error: option '-p, --port \u003Cnumber\u003E' argument 'aaaa' is invalid. not a valid integer.\n",
  "meta": {
    "expires": null
  }
}

exports[`parse stdout`] = {
  "data": "",
  "meta": {
    "expires": null
  }
}
